File: pe-core/src/main/java/org/saar/core/renderer/shadow/ShadowsRenderingPath.java
package org.saar.core.renderer.shadow;

import org.joml.Vector3fc;
import org.saar.core.camera.projection.OrthographicProjection;
import org.saar.core.light.IDirectionalLight;
import org.saar.core.renderer.RenderContextBase;
import org.saar.core.renderer.RenderingPath;
import org.saar.core.screen.OffScreen;
import org.saar.core.screen.Screens;
import org.saar.lwjgl.opengl.fbos.Fbo;
import org.saar.lwjgl.opengl.textures.TextureTarget;
import org.saar.lwjgl.opengl.textures.parameters.MagFilterParameter;
import org.saar.lwjgl.opengl.textures.parameters.MinFilterParameter;
import org.saar.lwjgl.opengl.textures.parameters.WrapParameter;
import org.saar.lwjgl.opengl.textures.settings.TextureMagFilterSetting;
import org.saar.lwjgl.opengl.textures.settings.TextureMinFilterSetting;
import org.saar.lwjgl.opengl.textures.settings.TextureSWrapSetting;
import org.saar.lwjgl.opengl.textures.settings.TextureTWrapSetting;
import org.saar.lwjgl.opengl.utils.GlBuffer;
import org.saar.lwjgl.opengl.utils.GlCullFace;
import org.saar.lwjgl.opengl.utils.GlUtils;

public class ShadowsRenderingPath implements RenderingPath {

    private final ShadowsScreenPrototype prototype;
    private final ShadowsCamera camera;
    private final OffScreen screen;

    private final ShadowsRenderNode renderNode;

    public ShadowsRenderingPath(ShadowsQuality quality, OrthographicProjection projection,
                                IDirectionalLight light, ShadowsRenderNode renderNode) {
        this(new ShadowsScreenPrototypeDefault(), quality, projection, light, renderNode);
    }

    public ShadowsRenderingPath(ShadowsScreenPrototype prototype, ShadowsQuality quality,
                                OrthographicProjection projection, IDirectionalLight light,
                                ShadowsRenderNode renderNode) {
        this.prototype = prototype;
        this.camera = camera(projection, light.getDirection());
        this.renderNode = renderNode;

        final Fbo fbo = Fbo.create(quality.getImageSize(), quality.getImageSize());
        this.screen = Screens.fromPrototype(this.prototype, fbo);
        this.prototype.getDepthTexture().setSettings(TextureTarget.TEXTURE_2D,
                new TextureMinFilterSetting(MinFilterParameter.LINEAR),
                new TextureMagFilterSetting(MagFilterParameter.LINEAR),
                new TextureSWrapSetting(WrapParameter.CLAMP_TO_EDGE),
                new TextureTWrapSetting(WrapParameter.CLAMP_TO_EDGE));
    }

    private static ShadowsCamera camera(OrthographicProjection projection, Vector3fc direction) {
        final ShadowsCamera shadowsCamera = new ShadowsCamera(projection);
        shadowsCamera.getTransform().getRotation().lookAlong(direction);
        shadowsCamera.getTransform().getPosition().set(0, 0, 0);
        return shadowsCamera;
    }

    public ShadowsCamera getCamera() {
        return this.camera;
    }

    @Override
    public ShadowRenderingOutput render() {
        this.screen.setAsDraw();
        GlUtils.clear(GlBuffer.DEPTH);

        final RenderContextBase context = new RenderContextBase(this.camera);
        context.getHints().cullFace = GlCullFace.FRONT;
        this.renderNode.renderShadows(context);

        return new ShadowRenderingOutput(this.screen, this.prototype.getDepthTexture());
    }

    @Override
    public void delete() {
        this.renderNode.delete();
        this.screen.delete();
    }
}

File: pe-core/src/main/java/org/saar/core/renderer/shadow/ShadowsRenderingPath.kt
package org.saar.core.renderer.shadow

import org.saar.core.camera.projection.OrthographicProjection
import org.saar.core.light.IDirectionalLight
import org.saar.core.renderer.RenderContext
import org.saar.core.renderer.RenderingOutput
import org.saar.core.renderer.RenderingPath
import org.saar.core.screen.Screens
import org.saar.lwjgl.opengl.constants.Face
import org.saar.lwjgl.opengl.cullface.CullFace
import org.saar.lwjgl.opengl.fbo.Fbo
import org.saar.lwjgl.opengl.fbo.attachment.allocation.SimpleAllocationStrategy
import org.saar.lwjgl.opengl.utils.GlBuffer
import org.saar.lwjgl.opengl.utils.GlUtils

class ShadowsRenderingPath(
    quality: ShadowsQuality,
    projection: OrthographicProjection,
    light: IDirectionalLight,
    private val renderNode: ShadowsRenderNode,
) : RenderingPath<ShadowsBuffers> {

    val camera: ShadowsCamera = ShadowsCamera(projection).apply {
        transform.rotation.lookAlong(light.direction)
        transform.position.set(0f, 0f, 0f)
    }

    private val prototype = ShadowsScreenPrototype()

    private val screen = Screens.fromPrototype(this.prototype,
        Fbo.create(quality.imageSize, quality.imageSize),
        SimpleAllocationStrategy())

    override fun render(): RenderingOutput<ShadowsBuffers> {
        this.screen.setAsDraw()

        GlUtils.clear(GlBuffer.DEPTH)

        val context = RenderContext(this.camera)

        CullFace.set(true, Face.FRONT)
        this.renderNode.renderShadows(context)

        return RenderingOutput(this.screen, this.prototype.buffers)
    }

    override fun delete() {
        this.renderNode.delete()
        this.screen.delete()
    }
}